### PR TITLE
feat(AWS Lambda): Add ability to customize file for Dockerfile

### PR DESCRIPTION
**Pull request by pgrzesik**
Closes: #8856

### CODE REVIEW

File: docs/providers/aws/guide/functions.md [hunk 1]
```diff
@@ -211,6 +211,7 @@ provider:
     images:
       baseimage:
         path: ./path/to/context
+        file: Dockerfile.dev
```
Comment (medikoo) [inline docs/providers/aws/guide/functions.md:214]: Does it make sense to have both `path` and `file`?

Wouldn't `file` make `path` obsolete? 

I wonder weather we could not rely just on `path`, and allow it to either point a directory that hosts `Dockerfile` or point `Dockerfile` specifically (?)

**Review by medikoo — COMMENTED** [review]
@pgrzesik great call on addressing that. Before I review it fully, I've posted one comment to clarify one thing (resolution on docker side is not perfectly clear to me)

File: docs/providers/aws/guide/functions.md (hunk 1 above)
Comment (pgrzesik) [reply docs/providers/aws/guide/functions.md:214]: I believe it does, for Docker `path` is translated to context `PATH` that is passed to `docker build`, where `file` is translated to `-f` which points to Dockerfile that needs to be used. By default, the `Dockerfile` is resolved with `path/Dockerfile`, but often `Dockerfiles` might reside in a separate directory, e.g. `docker`, where codebase is stored in `src`. In such case, `Dockerfile` won't be available in the provided context. In tools like `docker-compose`, these two are also separate options (`dockerfile` and `context`): https://docs.docker.com/compose/compose-file/compose-file-v3/#build  While in most cases we should be fine with just `file`, there are cases that require these two to be defined separately and for that reason, I would keep two separate options.

Comment (medikoo) [reply docs/providers/aws/guide/functions.md:214]: Thanks for clarifications indeed it makes sense. Now I think that we should have named `path` a `context` (or `contextPath`), at least it'll pair better with `file`, but let's leave it for now

Comment (pgrzesik) [reply docs/providers/aws/guide/functions.md:214]: I agree, `contextPath` would be probably the best name for that property

File: lib/plugins/aws/provider.js [hunk 2]
```diff
@@ -1894,11 +1895,15 @@ Object.defineProperties(
       { promise: true }
     ),
     resolveImageUriAndShaFromPath: d(
-      async function (imageName, imagePath) {
+      async function (imageName, imagePath, imageFilename = 'Dockerfile') {
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:1898]: Optional arguments are not taken into account in memoization, and I believe in that case this one should be considered.

To fix it we may either:
- Not make this argument optional (that's simplest), and always pass it to this function
- Make it optional, then we need to enforce `length: 3` setting, and ideally configure a _resolver_ for third argument, to ensure that not passing it is equal to passing `'Dockerfile'` (that can be done with option: `resolvers: [null, null, (imageFilename = 'Dockerfile') => imageFilename]`)

**Review by medikoo — CHANGES_REQUESTED** [review]
Thanks @pgrzesik, it looks very good, I've just spotted a likely issue in memoization setup.

File: lib/plugins/aws/provider.js (hunk 2 above)
Comment (pgrzesik) [reply lib/plugins/aws/provider.js]: Great catch, I was not aware of that :+1: I've changed the logic to always pass `imageFilename` here so memoization should work correctly :+1:

**Review by medikoo — APPROVED** [review]
Looks great 👍

**Pull request merged**